### ISSUE TITLE
Implement extended version of PSLQ (for Z[I]) and use that in field_isomorphism_pslq()

### ISSUE THREAD

**Issue by skirpichev** [diofant/diofant#929]
[Like Maple](https://www.maplesoft.com/support/help/Maple/view.aspx?path=IntegerRelations%2FPSLQ).  After finding a gaussian integer relation for powers of the primitive element, we can eliminate `I` using resultant.

For real algebraic fields, the difference is significant:
```python
In [1]: A = QQ.algebraic_field(3*sqrt(2)+2*sqrt(3)+1)

In [2]: B = QQ.algebraic_field(sqrt(2)+sqrt(3))

In [3]: from diofant.polys.numberfields import field_isomorphism

In [4]: %time field_isomorphism(A, B)
CPU times: user 32.2 ms, sys: 318 µs, total: 32.5 ms
Wall time: 42.9 ms
Out[4]: [1/2, 0, -5/2, 1]

In [5]: %time field_isomorphism(A, B, fast=False)
CPU times: user 1.03 s, sys: 4.68 ms, total: 1.04 s
Wall time: 1.06 s
Out[5]: [1/2, 0, -5/2, 1]
```

**Comment by skirpichev** [comment]
@jksuom, it seems for this case `Z[I]` would be really needed.  I doubt if they could be added to mpmath (after all, it seems unmaintained), so it may be better to implement `pslq()` in the diofant itself.  It could be possible to solve this issue without using gaussian integers (see an attempt in https://github.com/diofant/diofant/pull/945), but that seems less nice.

Are you planning to finish https://github.com/sympy/sympy/pull/15396?  If so, may be it does make sense to implement a generic algebraic integers classes (ring/element pair)?  Integral basis computation for algebraic field may be postponed, but for easy cases (quadratic fields?) we could do this right now easily.   Also, polys/solvers.py uses raw matrices, without `sympify()`, so it should be possible to use them for elements of the ring of integers.  Or do you think there are better variants for representation?

**Comment by jksuom** [comment]
Yes, I am planning to finish the gaussian domains PR (though there always seems to be something more important to do). It is special but probably useful in some situations. I think that there are several possible implementations for algebraic numbers (not only integers). It is possible that they should all be implemented together with suitable conversions methods.

**Comment by skirpichev** [comment]
> It is special but probably useful in some situations.

Your `QQ(I)`, probably, also will be little more efficient than the current gaussian rationals implementation.  But the GaussianRationalField must be subclassed from the AlgebraicField, isn't?

> It is possible that they should all be implemented together with suitable conversions methods.

Maybe, just like DMP and SMP (PolyElement) polynomials.

**Comment by jksuom** [comment]
Cohen (A Course in Computational Algebraic Number Theory) considers three types of representations.

1. A pair `(P, x)` where `P` is the minimal polynomial and `x` a complex approximation. This is essentially what `ComplexRootOf` does.

2. Given a generator `theta` of a number field, represent its numbers as polynomials in `theta`. Cohen suggests using a polynomial with integer coefficients and a common denominator. Presumably, this is more efficient that polynomials with rational coefficients.

3. Matrix of the regular representation. Again an integer matrix and a common denominator of all entries.

It looks like each one of these is useful for some tasks.

**Comment by skirpichev** [comment]
On Wed, Feb 19, 2020 at 03:41:09AM -0800, Kalevi Suominen wrote:
>     3. Matrix of the regular representation. Again an integer matrix and a
>        common denominator of all entries.

Only this seems acceptable in the case of the ring of integers
of some algebraic field.

**Comment by skirpichev** [comment]
I think, the 2-nd option does make sense, except that in general there is no power basis (in `theta`).

I'm not sure, however, that this representation will work for general finite extensions (Diofant has support for nested algebraic extensions).  E.g. suppose `K` is a field (including at least `QQ` and any it's finite extension), `t` - a primitive element for the field extension `K<t>` (degree `n`), the set `{t_k for k in range(n)}` - an integral basis.  Then any element of `K<t>` we can be represented as `sum(x_k*t_k/d for k in range(n))`, where `x_k` are from the ring of integers `O_K` and `d` is if from `O_K - {0}`. But can we make this representation effective and unique?  For `O_K==ZZ` we could make sure `d>0` and `gcd({x_k},d)==1`.  But in general, `O_K` will not be a Euclidean domain or even a UFD.

**Comment by skirpichev** [comment]
I'm start thinking that support for tower of extensions was a bad idea, it complicates things.  E.g. see above problem with a canonical representation for it's integers.  @jksuom, I would appreciate your opinion.  Maybe this problem could be solved for a special case `K=QQ<real>`?  But I doubt.

In fact, everything required for root isolation stuff: real & imag properties.  We could construct AlgebraicField's like `AlgebraicField(QQ, <real>)` if the polynomial has only real coefficients and `AlgebraicField(QQ, <real>, I)` if it has complex ones.  To catch the later case, we could (after finding its primitive element `theta`) just check if `I` and `theta.conjugate()` are in this field.  Given this, we can compute then `real` and `imag` parts for elements in this field (as elements of the real extension field `QQ<re(theta), im(theta)>`).

**Comment by jksuom** [comment]
> it complicates things

That is what I would have expected. An implementation with a single generator will have high degree but it might still be more efficient than a tower of low degree extensions.

> QQ<re(theta), im(theta)>

It looks like this could be an efficient way of embedding a number field into C (assuming that you can efficiently construct re(theta) and im(theta)).

**Comment by skirpichev** [comment]
> An implementation with a single generator will have high degree but it might still be more efficient

The problem rather with canonicalization (how to compute gcd, etc - to represent `O_{K<t>}` elements), not with efficiency.  I think, performance difference will be negligible, if any.

**Comment by jksuom** [comment]
> how to compute gcd

Are you thinking of gcd in something other than ZZ?

**Comment by skirpichev** [comment]
Yes.  Currently, we allow generic towers of extensions, i.e. finite extension over any algebraic field `K`.  For `K=QQ` it's possible to represent `K<t>` elements as `sum(x_k*t_k/d for k in range(n))`, where `x_k` and `d` are from `O_K == ZZ` (ring of integers) and `{t_k}` - an integral basis.   Here is gcd come to play, making this representation unique (assuming `d>0`).  But in general, `O_K` will not be even a UFD.

**Comment by jksuom** [comment]
But the integral basis should then also be canonical. Have you some plans for that?

**Comment by skirpichev** [comment]
What do you mean?  Any integral basis in `O_K<t>` should be fine.  By "unique" I mean the unique representation w.r.t to the given basis.

The computation of some integral basis is a different story (and I don't know how to do this without viewing `K<t>` as an absolute extension, i.e. over `QQ` - some `QQ<alpha>`).  I was planning to implement quadratic extensions first.  This will cover all interesting (so far) cases, i.e. the Gaussian integers - and integral basis computation here is easy.

**Comment by jksuom** [comment]
>  By "unique" I mean the unique representation w.r.t to the given basis.

Ok. Then there is a unique representation. But that representation may not be efficient for multiplications.

**Comment by skirpichev** [comment]
> Then there is a unique representation.

Hmm, I'm not sure (if `K!=QQ`).

**Comment by jksuom** [comment]
I was thinking of the case with an extension over K = QQ. Other cases where O_K would be a UFD are rare.
